upf/io/pddl_reader.py

```diff
@@ -14,24 +14,321 @@
+                       + ':duration' + nestedExpr().setResultsName('duration') \
+                       + ':condition' + nestedExpr().setResultsName('cond') \
+                       + ':effect' + nestedExpr().setResultsName('eff') \
+                       + Suppress(')'))
+
+pp_domain = Suppress('(') + 'define' + Suppress('(') + 'domain' + name + Suppress(')') \
+    + Optional(require_def) + Optional(types_def) + Optional(constants_def) \
+    + Optional(predicates_def) + Optional(functions_def) \
+    + Group(ZeroOrMore(action_def | dur_action_def)).setResultsName('actions') + Suppress(')')
+
+objects = OneOrMore(Group(Group(OneOrMore(name)) \
+                          + Optional(Suppress('-') + name))).setResultsName('objects')
+pp_problem = Suppress('(') + 'define' \
+    + Suppress('(') + 'problem' + name.setResultsName('name') + Suppress(')') \
+    + Suppress('(') + ':domain' + name + Suppress(')') + Optional(require_def) \
+    + Optional(Suppress('(') + ':objects' + objects + Suppress(')')) \
+    + Suppress('(') + ':init' + OneOrMore(nestedExpr()).setResultsName('init') + Suppress(')') \
+    + Suppress('(') + ':goal' + nestedExpr().setResultsName('goal') + Suppress(')') \
+    + Suppress(')')
+
+pp_domain.ignore(';' + restOfLine)
+pp_problem.ignore(';' + restOfLine)
 
 
 class PDDLReader:
     """
     Parse a PDDL problem and generate a upf problem.
-
-    The current implementation relies on tarski and has a bug parsing real
-    constants. See issue https://github.com/aig-upf/tarski/issues/114.
     """
     def __init__(self, env: Environment = None):
-        self.reader = tarski.io.PDDLReader(raise_on_error=True,
-                                           strict_with_requirements=False)
-        self.env = get_env(env)
-
-    def parse_problem(self, domain: str, problem: str) -> 'upf.model.Problem':
-        self.reader.parse_domain(domain)
-        problem = self.reader.parse_instance(problem)
-        return convert_tarski_problem(self.env, problem)
+        self._env = get_env(env)
+        self._em = self._env.expression_manager
+        self._tm = self._env.type_manager
+        self._operators = {
+            'and' : self._em.And,
+            'or' : self._em.Or,
+            'not' : self._em.Not,
+            'imply' : self._em.Implies,
+            '>=' : self._em.GE,
+            '<=' : self._em.LE,
+            '>' : self._em.GT,
+            '<' : self._em.LT,
+            '=' : self._em.Equals,
+            '+' : self._em.Plus,
+            '-' : self._em.Minus,
+            '/' : self._em.Div,
+            '*' : self._em.Times
+        }
+
+    def _parse_exp(self, problem, act, var, exp):
+        if isinstance(exp, ParseResults) and len(exp) == 0:
+            return self._em.TRUE()
+        elif isinstance(exp, ParseResults) and exp[0] == '-' and len(exp) == 2:
+            return Times(-1, self._parse_exp(problem, act, var, exp[1]))
```

Comment:
Maybe remove recursion here: it should be a matter of using a stack and a map for the results, no?